src/plugins/ui_feature/find_and_replace.ts

```diff
@@ -138,35 +150,93 @@ export class FindAndReplacePlugin extends UIPlugin {
     }
     this.currentSearchRegex = RegExp(searchValue, flags);
   }
+
+  private getSearchSheets() {
+    let sheets: UID[] = [];
+    switch (this.searchOptions.searchScope) {
+      case "allSheets":
+        sheets = this.getters.getSheetIds();
+        const activeSheetIndex = sheets.findIndex((id) => id === this.getters.getActiveSheetId());
+        sheets = [
+          sheets[activeSheetIndex],
+          ...sheets.slice(activeSheetIndex + 1),
+          ...sheets.slice(0, activeSheetIndex),
+        ];
+        break;
+      case "thisSheet":
+        sheets = [this.getters.getActiveSheetId()];
+        break;
+      case "specificRange":
+        const specificRange = this.searchOptions.specificRange;
+        if (!specificRange) {
+          return [];
+        }
+        sheets = [specificRange.sheetId];
+        break;
+    }
+    return sheets;
+  }
+
   /**
    * Find matches using the current regex
    */
   private findMatches() {
-    const sheetId = this.getters.getActiveSheetId();
-    const cells = this.getters.getCells(sheetId);
     const matches: SearchMatch[] = [];
+    let sheets: UID[];
+    const specificRange = this.searchOptions.specificRange;
+    if (!this.toSearch) {
+      return matches;
+    }
+
+    sheets = this.getSearchSheets();
+
+    for (const sheetId of sheets) {
+      matches.push(...this.findMatchesInSheet(sheetId, specificRange));
+    }
+    matches.sort((a, b) => this.sortBySheetinitialStateIndexThenRowThenColumn(a, b));
+
+    const activeMatchIndex = matches.findIndex(
+      (match) => match.sheetId === this.getters.getActiveSheetId()
+    );
+    this.selectedMatchIndex = null;
+    this.selectedMatchOffset = activeMatchIndex;
+
+    return matches;
+  }
+
+  private findMatchesInSheet(sheetId: string, specificRange?: Range) {
+    const matches: SearchMatch[] = [];
+    const rangeData =
+      this.searchOptions.searchScope === "specificRange" && specificRange
+        ? specificRange.zone
+        : this.getters.getSheetZone(sheetId);
+
+    if (!this.currentSearchRegex) {
+      return [];
```

Comment:
We assign currentSearchRegex as null in the initial state and also when clearing the search. Therefore, we receive a TypeScript error indicating that it can be null. To resolve this, we can initialize it with `new RegExp("")`, and when clearing the search, we can set the same value to currentSearchRegex. By doing so, we can eliminate the need for this condition

Reply:
you ddon't understand. this is what I meant https://github.com/odoo/o-spreadsheet/pull/2330/files#r1229381286